### PR TITLE
[CIT-386] Mapper tests use systemd services

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes

- Fix the existing mapper tests to use systemd services instead of direct uses of binaries
- Add more tests using c8y mapper and az mapper

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+from pysys.basetest import BaseTest
```
Comment (PradeepKiruvale) [inline tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py:1]: Do we need one such test for az mapper as well?

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            arguments=["systemctl", "start", "tedge-mapper-c8y"],
+            stdouterr="tedge_mapper_c8y",
+        )
+
+        self.addCleanupFunction(self.mapper_cleanup)
+
+    def execute(self):
+        sub = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "sub", "--no-topic", "c8y/measurement/measurements/create"],
```
Comment (PradeepKiruvale) [inline tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py:37]: Why specifically this c8y/measurement/measurements/create topic?

**Review by PradeepKiruvale — COMMENTED** [review]
Code changes LGTM, added some comments.

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_negative/run.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+from pysys.basetest import BaseTest
+
+import time
+
+"""
+Validate an invalid Thin Edge JSON message that is published
```
[the remark this answers is not in the document]
Comment (rina23q) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_negative/run.py:6]: Invalid Thin Edge JSON, however, the test case catches the error of invalid JSON. Should I add the case to publish a message which is valid JSON but invalid Thin Edge JSON? 

e.g.
```
{
   "type": 12,
}
```

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py (hunk 1 above)
Comment (rina23q) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py:1]: I added here two positive test cases for az. Maybe I didn't get what you mean.

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py (hunk 2 above)
Comment (rina23q) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py:37]: c8y mapper publishes on `c8y/measurement/measurements/create`.

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_without_timestamp/run.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+Given a configured system
+When we set az.mapper.timestamp false
+When we start the tedge-mapper-az systemctl service
+When we start tedge sub with sudo in the background
+When we publish a Thin Edge JSON message
+Then we kill tedge sub with sudo as it is running with a different user account
+Then we validate the JSON message in the output of tedge sub
+Then we stop the tedge-mapper-az systemctl service
+Then we unset az.mapper.timestamp
```
Comment (abelikt) [inline tests/PySys/tedge_mapper_az/tedge_mapper_az_without_timestamp/run.py:18]: Oh, that is new to me. Do we maybe have a list of configuration options somewhere?

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py (hunk 1 above)
Comment (abelikt) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py:1]: Better to have them. Lets do that when the mapper testbed is created, there it should be easier to achieve.

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py (hunk 2 above)
Comment (abelikt) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_positive/run.py:37]: That is the topic (topic template (?)) where all the other c8y measurements go to.

**Comment by abelikt** [comment]
Looks good. All these tests still look complex. I'm currently working on a concept for the mapper testbed. It should be far more easier there.

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_without_timestamp/run.py (hunk 4 above)
Comment (rina23q) [reply tests/PySys/tedge_mapper_az/tedge_mapper_az_without_timestamp/run.py:18]: The output of `tedge config list --doc` is pretty much that you wanted.
`az.mapper.timestamp` is documented in https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/architecture/mapper.md#configure-whether-adding-a-timestamp-or-not.

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_negative/run.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        # Wait for a small amount of time to give tedge sub time
+        # to initialize. This is a heuristic measure.
+        # Without an additional wait we observe failures in 1% of the test
+        # runs.
+        time.sleep(0.1)
+
+        pub = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "pub",
+                       "tedge/measurements", '{"temperature": 12,}'],
```
[the remark this answers is not in the document]
Comment (rina23q) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_negative/run.py:50]: In theory, only `{` and `{"temperature": 12,}` can be considered same as both are wrongly formatted JSON. I am not sure that I should add a case it is a valid JSON but invalid Thin Edge JSON. For example,
`{"type": "aaaa"}`

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_negative/run.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        # Wait for a small amount of time to give tedge sub time
+        # to initialize. This is a heuristic measure.
+        # Without an additional wait we observe failures in 1% of the test
+        # runs.
+        time.sleep(0.1)
+
+        pub = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "pub",
+                       "tedge/measurements", '{"temperature": 12},'],
```
[the remark this answers is not in the document]
Comment (rina23q) [reply tests/PySys/tedge_mapper_az/tedge_mapper_az_negative/run.py:49]: I change to `{`, it doesn't matter if it is an invalid JSON

File: tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_negative/run.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rina23q) [reply tests/PySys/tedge_mapper_c8y/tedge_mapper_c8y_negative/run.py]: Decided to change it to `an invalid JSON` as the purpose of test is to confirm error messages are really published on `tedge/errors`. The message doesn't matter.

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_negative/pysystest.xml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<pysystest type="auto">
+	<description>
+		<title>
+			Validate the Azure Mapper publishes an error message on tedge/errors
+			when it receives an invalid JSON message on tedge/measurements.
```
Comment (PradeepKiruvale) [inline tests/PySys/tedge_mapper_az/tedge_mapper_az_negative/pysystest.xml:6]: Azure Mapper that publishes

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_with_timestamp/pysystest.xml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="utf-8"?>
+<pysystest type="auto">
+	<description>
+		<title>
+			Validate the Azure Mapper publishes a Thin Edge JSON message
```
Comment (PradeepKiruvale) [inline tests/PySys/tedge_mapper_az/tedge_mapper_az_with_timestamp/pysystest.xml:5]: Azure Mapper that publishes?

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_without_timestamp/pysystest.xml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<pysystest type="auto">
+	<description>
+		<title>
+			Validate the Azure Mapper publishes a Thin Edge JSON message
```
Comment (PradeepKiruvale) [inline tests/PySys/tedge_mapper_az/tedge_mapper_az_without_timestamp/pysystest.xml:5]: Same as above

**Review by PradeepKiruvale — APPROVED** [review]

**Review by abelikt — APPROVED** [review]
Looks good and passes all tests

File: tests/PySys/tedge_mapper_az/tedge_mapper_az_negative/pysystest.xml (hunk 7 above)
Comment (rina23q) [reply tests/PySys/tedge_mapper_az/tedge_mapper_az_negative/pysystest.xml:6]: Added `that` to all :D

**Pull request merged**